### PR TITLE
Adds exception to handle nil buildfile issue

### PR DESCRIPTION
**Pull request by sachin-sandhu**
### What are you trying to accomplish?

Fixes following issue:

`NoMethodError: undefined method 'content' for nil:NilClass`

```
  original_content = buildfile.content.dup
                              ^^^^^^^^
  gradle/lib/dependabot/gradle/file_updater.rb:116:in `update_version_in_buildfile'
  gradle/lib/dependabot/gradle/file_updater.rb:74:in `block in update_buildfiles_for_dependency'
  gradle/lib/dependabot/gradle/file_updater.rb:62:in `each'
  gradle/lib/dependabot/gradle/file_updater.rb:62:in `update_buildfiles_for_dependency'
  gradle/lib/dependabot/gradle/file_updater.rb:28:in `block in updated_dependency_files'
...
(27 additional frame(s) were not displayed)
```

**Preface:** 

When managing a large project with several sub-modules, user can work with different configurations. user can create a single repo with all the submodules as part of single project. In this way, all sub modules can be handled as part of single repository.

Another approach would be to manage the submodules in separate repos. Submodules can be added to a parent project via `git submodule add https://github.com/<user>/repo repo` This way each project (main and sub modules) can be managed differently from each other, this is helpful in cases when a project is intended to be used a sub module in several projects. an example snapshot of a sub module is as following. Here, **blockmiui** and **xtoast** are added as submodules to a project. These submodules are managed from their own separate repos.

image

**Issue:** This issue exists when :dependabot: makes an update to a _gradle_ project. It is observed that dependabot is parsing the files in submodules and treating the files as part of a single repo. This is evident when file_fetcher also treats files ( build.gradle, build.gradle.kts ) from submodules eligible for update. Incidentally, when the file_updater module runs, the submodule project files are not parsed. So in this scenario, a dependency from a submodule while found to be eligible to update, does not have a valid file where it can be updated as submodule files are filtered out in file_updater stage. So no relevant build file is found at `original_content = buildfile.content.dup` in `update_version_in_buildfile`. 

**Fix:** Raise "No files changed!" exception if file_updater is not able to find a valid build file eligible for update.

### How will you know you've accomplished your goal?

If Build file update is not successful, a new exception with message "No files changed!" will be raised.

### Checklist

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

### CODE REVIEW

File: gradle/lib/dependabot/gradle/file_updater.rb [hunk 1]
```diff
@@ -69,6 +69,13 @@ def update_buildfiles_for_dependency(buildfiles:, dependency:)
 
           buildfile = files.find { |f| f.name == new_req.fetch(:file) }
 
+          # Exception raised to handle issue that arises when buildfiles function (see this file)
+          # removes the build file that contains the dependency itself. So no build file exists to
+          # update dependency, This behaviour is evident for only extremely small number of users
+          # that have added separate repos as sub-modules in parent projects
+
+          raise "No files changed!" if buildfile.nil?
```
Comment (abdulapopoola) [inline gradle/lib/dependabot/gradle/file_updater.rb:77]: can we have a more specific exception class please?

**Comment by abdulapopoola** [comment]
Is this happening because we don't have _sub-modules_ support?

**Comment by sachin-sandhu** [comment]
> Is this happening because we don't have _sub-modules_ support?

I could not confirm it because :dependabot: is parsing the target repo (with sub modules) as one repo in file_fetcher stage. So while making list of dependencies, it is also taking the sub module files into account. But when running the file_updater, :dependabot: filters out the sub module folders from list. But in either of cases, sub modules should have their own dependency management.

File: gradle/lib/dependabot/gradle/file_updater.rb [hunk 2]
```diff
@@ -69,6 +80,18 @@ def update_buildfiles_for_dependency(buildfiles:, dependency:)
 
           buildfile = files.find { |f| f.name == new_req.fetch(:file) }
 
+          # Exception raised to handle issue that arises when buildfiles function (see this file)
+          # removes the build file that contains the dependency itself. So no build file exists to
+          # update dependency, This behaviour is evident for extremely small number of users
+          # that have added separate repos as sub-modules in parent projects
+
+          if buildfile.nil?
```
Comment (abdulapopoola) [inline gradle/lib/dependabot/gradle/file_updater.rb:88]: can we have test coverage too please?

File: gradle/lib/dependabot/gradle/file_updater.rb [hunk 3]
```diff
@@ -12,6 +12,17 @@ module Gradle
...
+      class BuildFileNotFound < StandardError
+        def initialize(message:, error_context:)
+          super(message)
+          @error_context = error_context
+        end
+
+        def sentry_context
+          { extra: @error_context }
+        end
+      end
```
Comment (jeffwidman) [inline gradle/lib/dependabot/gradle/file_updater.rb:24]: @abdulapopoola is right, this shouldn't be a one-off error... this type of problem is actually a common class of errors across all our package managers where we can't find the dependency file.

Our well known errors are here: https://github.com/dependabot/dependabot-core/blob/ae1e814b35ce06d9d36c210040182250ba22a05a/common/lib/dependabot/errors.rb

It's been a minute since I looked at this, but a quick skim shows separate sections for File Fetcher, File Parser, and File Updater... given this is in the file updater, I suspect you want this section: https://github.com/dependabot/dependabot-core/blob/ae1e814b35ce06d9d36c210040182250ba22a05a/common/lib/dependabot/errors.rb#L162

I don't know if you want `DependencyFileNotResolvable` or `GitDependenciesNotReachable` or `GitDependencyReferenceNotFound`... 

I suggest grep'ing the codebase for other examples of where these exceptions are used to understand which might be the most appropriate.

(I actually wonder if `GitDependencyReferenceNotFound` and `GitDependenciesNotReachable` should be collapsed, but that's a separate conversation).

File: gradle/lib/dependabot/gradle/file_updater.rb (hunk 2 above)
Comment (jeffwidman) [reply gradle/lib/dependabot/gradle/file_updater.rb:88]: Should be straightforward to add an example manifest to test, if you run into problems just ask.

**Review by jeffwidman — CHANGES_REQUESTED** [review]
Let's re-use one of our pre-existing well-known errors unless you have a compelling reason why this is different.

**Comment by jeffwidman** [comment]
Also, great description of the underlying problem... thanks for including the screenshots and code notes, really helped me grok it.

File: gradle/lib/dependabot/gradle/file_updater.rb (hunk 2 above)
Comment (sachin-sandhu) [reply gradle/lib/dependabot/gradle/file_updater.rb]: Added test case to mock non existent dependency file. Please check and let know if that suffices.

File: gradle/lib/dependabot/gradle/file_updater.rb (hunk 3 above)
Comment (sachin-sandhu) [reply gradle/lib/dependabot/gradle/file_updater.rb]: DependencyFileNotResolvable seems a good error class to me

**Comment by sachin-sandhu** [comment]
@jeffwidman , Please review once you have some time today , thanks !

**Comment by sachin-sandhu** [comment]
@jeffwidman , please review once you are available , thanks !

File: gradle/lib/dependabot/gradle/file_updater.rb [hunk 4]
```diff
@@ -69,6 +68,13 @@ def update_buildfiles_for_dependency(buildfiles:, dependency:)
 
           buildfile = files.find { |f| f.name == new_req.fetch(:file) }
 
+          # Exception raised to handle issue that arises when buildfiles function (see this file)
+          # removes the build file that contains the dependency itself. So no build file exists to
+          # update dependency, This behaviour is evident for extremely small number of users
+          # that have added separate repos as sub-modules in parent projects
+
+          raise DependencyFileNotResolvable, "No build file found to update the dependency" if buildfile.nil?
```
[the remark this answers is not in the document]
Comment (sachin-sandhu) [reply gradle/lib/dependabot/gradle/file_updater.rb:76]: @abdulapopoola , Let me check on this and come back to you.

[the remark this answers is not in the document]
Comment (sachin-sandhu) [reply gradle/lib/dependabot/gradle/file_updater.rb:76]: @abdulapopoola ,
This issue is specific to Gradle only because Sentry logged it considering the error context (throw position, variable context)
I need to check what approach do we have for sub modules. Might be only Gradle parser is having this issue. Haven't checked for other eco systems. Let me check on it and come back to you.

[the remark this answers is not in the document]
Comment (abdulapopoola) [reply gradle/lib/dependabot/gradle/file_updater.rb:76]: @sachin-sandhu , you're right, Sentry also only shows [gradle](https://github.sentry.io/discover/results/?field=title&field=release&field=environment&field=user.display&field=timestamp&field=gh.dependabot_api.update_config.package_manager&name=NoMethodError%3A+undefined+method+%60content%27+for+nil&project=1451818&query=issue%3ADELTAFORCE-Z59+%21gh.dependabot_api.update_config.package_manager%3A%22%22&sort=gh.dependabot_api.update_config.package_manager&statsPeriod=90d&yAxis=count%28%29).

It does seem that the error has stopped now but it's still good to add in the more specific error so that if it recurs, we can quickly root cause it.

[the remark this answers is not in the document]
Comment (sachin-sandhu) [reply gradle/lib/dependabot/gradle/file_updater.rb:76]: @abdulapopoola , Gave it another thought. I think instead of throwing exception which will end up in Sentry again, Another way would be to just exclude the update for particular dependency and let update continue for others.

[the remark this answers is not in the document]
Comment (abdulapopoola) [reply gradle/lib/dependabot/gradle/file_updater.rb:76]: Makes a lot of sense since we do a best effort approach, I would hope we write the logs about why this update fails (e.g. for grouped update scenarios). I also still think we should have a proper error class for this (e.g. SubModulesNotConfigured or something similar).

[the remark this answers is not in the document]
Comment (jeffwidman) [reply gradle/lib/dependabot/gradle/file_updater.rb:76]: > Another way would be to just exclude the update for particular dependency and let update continue for others.

It's a nice idea, but if we can't pull the dependency, than how can we safely walk the tree? Like the unreachable dependency may have a version pin on another dependency in the tree that prevents updating it.

I know we've had similar requests in Core of "please ignore my private deps that are on some private registry" and we invariably end up realizing without the full tree we can't propose any updates with confidence.

File: gradle/lib/dependabot/gradle/file_updater.rb [hunk 5]
```diff
@@ -69,6 +68,13 @@ def update_buildfiles_for_dependency(buildfiles:, dependency:)
 
           buildfile = files.find { |f| f.name == new_req.fetch(:file) }
 
+          # Exception raised to handle issue that arises when buildfiles function (see this file)
+          # removes the build file that contains the dependency itself. So no build file exists to
+          # update dependency, This behaviour is evident for extremely small number of users
+          # that have added separate repos as sub-modules in parent projects
+
```
Comment (jeffwidman) [inline gradle/lib/dependabot/gradle/file_updater.rb:75]: I found this comment a little confusing, so can we reword it to provide better context to the reader?

I took a stab at it, but I'm not 100% sure I fully understand the issue, so if I misstated something, please correct my wording

> Currently, Dependabot assumes that Gradle projects using Gradle submodules are all in a single repo. However, some projects are actually using git submodule references for the Gradle submodules. When this happens, Dependabot's FileFetcher thinks the Gradle submodules are eligible for update, but then the FileUpdater filters out the git submodule reference from the build file. So we end up with no relevant build file, leaving us with no way to update that dependency.
> TODO: Figure out a way to actually navigate this rather than throwing an exception.

File: gradle/lib/dependabot/gradle/file_updater.rb (hunk 4 above)
Comment (jeffwidman) [inline gradle/lib/dependabot/gradle/file_updater.rb:76]: Also, I really like your wording of this exception message because it describes what happened, but not assuming how we got into this situation... 

The code comment is helpful to explain one way we get into this situation, but there may be more ways... so keeping the actual exception message pretty generic like you did seems best. 👍

**Review by jeffwidman — APPROVED** [review]
This looks good to me, other than I'd probably tweak the code comment for clarity...

I took a stab at it, but I'm not sure I 100% understand the problem, so feel free to say it differently.

File: gradle/lib/dependabot/gradle/file_updater.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeffwidman) [reply gradle/lib/dependabot/gradle/file_updater.rb:76]: Also, since `DependencyFileNotResolvable` is a well-known error which we classify as "problem with the user's repo" and not "our internal code error", it won't show up in Sentry again for us...

Here it is arguably a code issue with Dependabot, although I'd say it's an edge case missing feature rather than an out and out bug...
